src/axom/multimat/multimat.hpp

```diff
@@ -0,0 +1,765 @@
+   * each material, or for each cell x material.
+   *
+   * \param field_idx the index of the field
+   */
+  FieldMapping getFieldMapping(int field_idx) const
+  {
+    return m_fieldMappingVec[field_idx];
+  }
+
+  //Dynamic mode functions
+
+  /** Convert the layout to use dynamic layout. In dynamic layout, users can
+   *  call addEntry() or removeEntry() to change the materials in a cell.
+   */
+  void convertToDynamic();
+  /** Convert the layout to use static layout. **/
+  void convertToStatic();
+
+  /**
+   * \brief Add a material in a cell.
+   *
+   * In a cell-dominant layout, firstIdx is the cell index, secondIdx is the
+   * material index. And it's reverse for material-dominant layout.
+   *
+   * \return true if the material is added in the. False if it already exists.
+   */
+  bool addEntry(int firstIdx, int secondIdx);
+  /**
+   * \brief Remove a material from a cell
+   *
+   * In a cell-dominant layout, firstIdx is the cell index, secondIdx is the
+   * material index. And it's reverse for material-dominant layout.
+   *
+   * \return true if a material is removed. False if it was not in the cell.
+   */
+  bool removeEntry(int firstIdx, int secondIdx);
+
+  /** Print the detail of this object */
+  void print() const;
+  /**
+   * \brief  Return true if the object is valid, false otherwise.
+   *
+   * \param verboseOutput (Optional) if true, this function will print out its
+   * valid/invalid detail.
+   */
+  bool isValid(bool verboseOutput = false) const;
+
+protected:
+  //Return the Set pointer associalted with the given FieldMapping or field idx
+  SetType* get_mapped_set(FieldMapping);
+  SetType* get_mapped_set(int field_idx);
+  //Return the BivariateSet used for the specified layouts or the field
+  BivariateSetType* get_mapped_biSet(DataLayout, SparsityLayout);
+  BivariateSetType* get_mapped_biSet(int field_idx);
+  //Return the relation for the specified field idx or the mapping
+  StaticVariableRelationType* getRel(int field_idx);
+
+private:  //private functions
+  //Given a relation (cell->mat or mat->cell), create the other relation
+  void makeOtherRelation();
+
+  //helper functions
+  template <typename DataType>
+  void convertToSparse_helper(int map_i);
+  template <typename DataType>
+  void convertToDense_helper(int map_i);
+
+  template <typename DataType>
+  void transposeField_helper(int field_idx);
+
+  template <typename T>
+  MapBaseType* helper_copyField(const MultiMat&, int map_i);
+
+private:
+  unsigned int m_ncells, m_nmats;
+
+  //slam set variables
+  RangeSetType m_cellSet;
+  RangeSetType m_matSet;
+  //slam relation variables (for sparse-layout fields)
+  //Depending on the layout of the field, each field can be mapped to different Relations
+  //Relation can be nullptr if no field is using said relation
+  //cell to mat relations
+  std::vector<SetPosType> m_cellMatRel_beginsVec;
+  std::vector<SetPosType> m_cellMatRel_indicesVec;
```

Comment:
Still to be SLAM-ified?

Reply:
SLAM-ed? SLAM-erated?

Reply:
SLAM is designed not to own any data - these are probably the backing `std::vectors` for our relations.